src/libcore/ops/drop.rs

```diff
@@ -1,85 +1,120 @@
-/// then `Dropping Inner!`.
-///
-/// ```
-/// struct Inner;
-/// struct Outer(Inner);
-///
-/// impl Drop for Inner {
-///     fn drop(&mut self) {
-///         println!("Dropping Inner!");
-///     }
+/// 
+/// Here, `main` will print `Dropping!`. In other words, the compiler generates code that
+/// kind of looks like this:
+/// 
+/// ```rust,ignore
+/// fn main() {
+///     let mut _x = HasDrop;
+/// 
+///     Drop::drop(&mut x);
 /// }
-///
-/// impl Drop for Outer {
+/// ```
+/// 
+/// As you can see, our custom implementation of `Drop` will be called at the end of `main`.
+/// 
+/// ## You cannot call `Drop::drop` yourself
+/// 
+/// Because the compiler automatically calls `Drop::drop`, you cannot call it yourself. This
+/// would lead to "double drop", where `Drop::drop` is called twice on the same value. This
+/// can lead to things like "double frees".
+/// 
+/// In other words, if you tried to write the code in the previous example with the explicit
+/// call to `Drop::drop`, you'd get a compiler error.
+/// 
+/// If you'd like to call `drop` yourself, there is something you can do: call [`std::mem::drop`].
+/// This function will drop its argument. For more, see its documentation.
+/// 
+/// [`std::mem::drop`]: ../../std/mem/fn.drop.html
+/// 
+/// ## `Drop` is recursive
+/// 
+/// If your type is something like a `struct` or `enum` that is an aggregate of other types, then
+/// Rust will call `Drop::drop` on the type first, and then recursively on everything it contains.
+/// For example:
+/// 
+/// ```rust
+/// struct HasDrop;
+/// 
+/// impl Drop for HasDrop {
 ///     fn drop(&mut self) {
-///         println!("Dropping Outer!");
+///         println!("Dropping HasDrop!");
 ///     }
 /// }
-///
-/// fn main() {
-///     let _x = Outer(Inner);
+/// 
+/// struct HasTwoDrops {
+///     one: HasDrop,
+///     two: HasDrop,
 /// }
-/// ```
-///
-/// ## Variables are dropped in reverse order of declaration
-///
-/// `_first` is declared first and `_second` is declared second, so `main` will
-/// print `Declared second!` and then `Declared first!`.
-///
-/// ```
-/// struct PrintOnDrop(&'static str);
-///
-/// impl Drop for PrintOnDrop {
+/// 
+/// impl Drop for HasTwoDrops {
 ///     fn drop(&mut self) {
-///         println!("{}", self.0);
+///         println!("Dropping HasTwoDrops!");
 ///     }
 /// }
-///
+/// 
 /// fn main() {
-///     let _first = PrintOnDrop("Declared first!");
-///     let _second = PrintOnDrop("Declared second!");
+///     let _x = HasTwoDrops { one: HasDrop, two: HasDrop };
 /// }
 /// ```
+/// 
+/// This will print
+/// 
+/// ```text
+/// Dropping HasTwoDrops!
+/// Dropping HasDrop!
+/// Dropping HasDrop!
+/// ```
+/// 
+/// Similarly, a slice will drop each element in order.
+/// 
+/// Which of our two `HasDrop` drops first, though? It's the same order that
+/// they're declared: first `one`, then `two`. If you'd like to try this
+/// yourself, you can modify `HasDrop` above to contain some data, like an
```

Comment:
> you can modify `HasDrop` above to contain some data

Should we proactively do that to prove the point?

Reply:
I was thinking about it but 1. it's long 2. it's not a huge change, and people sometimes like examples they can try. i'm open to including it, it'd just make this section even longer than it already is. maybe that's okay!